testing/demo/pkg/transfer/relay.go

```diff
@@ -0,0 +1,261 @@
+package main
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+
+	proverclient "github.com/celestiaorg/celestia-zkevm-ibc-demo/provers/client"
+	"github.com/celestiaorg/celestia-zkevm-ibc-demo/testing/demo/pkg/ethereum"
+	"github.com/celestiaorg/celestia-zkevm-ibc-demo/testing/demo/pkg/utils"
+	coretypes "github.com/cometbft/cometbft/rpc/core/types"
+	transfertypes "github.com/cosmos/ibc-go/v10/modules/apps/transfer/types"
+	"github.com/cosmos/solidity-ibc-eureka/abigen/ics26router"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// FungibleTokenPacketData represents the ICS20 token transfer data structure
+// It should match IICS20TransferMsgs.FungibleTokenPacketData in the Solidity contract
+type FungibleTokenPacketData struct {
+	Denom    string
+	Amount   *big.Int
+	Sender   string
+	Receiver string
+	Memo     string
+}
+
+// relayByTx implements the logic of an IBC relayer.
+// It processes source tx, extracts IBC events, generates proofs,
+// and creates an Ethereum transaction to submit to the ICS26Router contract.
+func relayByTx(sourceTxHash string, targetClientID string) error {
+	fmt.Printf("Relaying IBC transaction %s to client %s...\n", sourceTxHash, targetClientID)
+
+	event, err := getSendPacketEvent(sourceTxHash)
+	if err != nil {
+		return fmt.Errorf("failed to get SendPacket event: %w", err)
+	}
+	fmt.Printf("SendPacket event: %v\n", event)
+
+	resp, err := getCelestiaProverResponse(event)
```

Comment:
I would mention this is the state membership proof not the STF.

Also I feel like for safety, we should pass in the latest app hash we updated the tendermint client to to make sure that the membership proof is with reference to that state root. I think it's fine now because no other transactions are submitted but I worry due to the lack of atomicity that this may generate a membership proof up to a different state root

Reply:
Ah I see that the prover itself handles this by having the client ID passed through in the request

Reply:
Yep exactly. It was difficult to query the last trusted height on the EVM contract from Go code but it's easy to do in Rust b/c there are wrappers. So I opted to just move it inside the celestia-prover.